WalletWasabi/Crypto/GroupElement.cs

```diff
@@ -28,6 +28,11 @@ public GroupElement(GEJ groupElement)
 
 		public static GroupElement Infinity { get; } = new GroupElement(GE.Infinity);
 
+		/// <summary>
+		/// The base point defined in the secp256k1 standard used in ECDSA public key derivation.
+		/// </summary>
+		public static GroupElement StandardGenerator { get; } = new GroupElement(EC.G);
```

Comment:
All curves have a generator called "G" (I don't know how they are chosen) but when you see `G` you know what that point is however if you see StandardGenerator you don't know what point is that. You can have a `secp256k1_curve_generator` that is perfectly clear but too long so, I would prefer `G`.

Reply:
@nothingmuch ?

Reply:
Let's go with G and for other generators we could use our notations used in the WabiSabi paper.

Reply:
> Let's go with G

https://www.youtube.com/watch?v=umDr0mPuyQc

Reply:
If we call `G` as `StandardGenerator` then how would we call `Gh`? `WabiSabiGeneratorForPedersenBlindingFactor`?

Reply:
I think the `Wasabi` prefix isn't necessary as it can be inferred from the context or just added as a comment that this is something custom.

Reply:
i think G should refer to the generator defined in the secp256k1 spec to avoid confusion... this is also why i suggested "standard generator" to refer to it.

but what should we use it for? Our `G_g` generator? `G_h`? unlike CT there's no real benefit for compatibility, unless we want Pedersen commitments for amounts to be compatible with libsecp256k1-zkp (the only reason i can think of is slightly reducing the burden on alternative implementations), in which case it should be used as `G_h` and `G_g` should be defined using that sha-256 hashing.

A simpler approach is to ignore that generator entirely, it is pretty arbitrary after all, and just generate *all* the generators we need as `HashToG(generator_name)`.

Reply:
That's an excellent point. We don't need that point at all. 

I am repeating myself again and again but I think this is what we need, just rename `GroupElementFromText` to `HashToG`:

```c#
// Generators for MAC and Show
public static readonly GroupElement Gw = GroupElementFromText(nameof(Gw));
public static readonly GroupElement Gwp = GroupElementFromText(nameof(Gwp));
public static readonly GroupElement Gx0 = GroupElementFromText(nameof(Gx0));
public static readonly GroupElement Gx1 = GroupElementFromText(nameof(Gx1));
public static readonly GroupElement GV = GroupElementFromText(nameof(GV));

// Generators for Pedersen commitments
public static readonly GroupElement Gg = GroupElementFromText(nameof(Gg));
public static readonly GroupElement Gh = GroupElementFromText(nameof(Gh));

// Generator for attributes M_{ai}
public static readonly GroupElement Ga = GroupElementFromText(nameof(Ga));

// Generator for serial numbers 
public static readonly GroupElement Gs = GroupElementFromText(nameof(Gs));
```

Reply:
this makes sense to me, or `HashToCurve`, since `HashToG` in the CPZ19 paper makes it clear that the G being referred there is the group itself, since it's typeset as ![](https://render.githubusercontent.com/render/math?math=\mathrm{HashTo}\mathbb{G}).

i think i personally still prefer if G_g and G_h were defined as in libsecp256k1-zkp, and the assumption of independence still works out if the definition of the hash to curve function also commits to the standard base point, IOW:

 ```cs
// Generators for MAC and Show
public static readonly GroupElement Gw = HashToCurve(EC.G.ToC() + nameof(Gw));
public static readonly GroupElement Gwp = HashToCurve(EC.G.ToC() + nameof(Gwp));
public static readonly GroupElement Gx0 = HashToCurve(EC.G.ToC() + nameof(Gx0));
public static readonly GroupElement Gx1 = HashToCurve(EC.G.ToC() + nameof(Gx1));
public static readonly GroupElement GV = HashToCurve(EC.G.ToC() + nameof(GV));

// Generators for Pedersen commitments
public static readonly GroupElement Gg = LegacyHashToCurve(EC.G.ToC());
public static readonly GroupElement Gh = EC.G

// Generator for attributes M_{ai}
public static readonly GroupElement Ga = HashToCurve(EC.G.ToC() + nameof(Ga));

// Generator for serial numbers 
public static readonly GroupElement Gs = HashToCurve(EC.G.ToC() + nameof(Gs));
```


Above, please pretend that `ge.ToC()` returns the uncompressed representation (`0x04` followed by just the full x and y coords, 32 bytes each, instead of pretty printing, which is what `ToC` actually does. I couldn't actually find such an encoding in NBitcoin's code if it exists, only `TryParse`.

`LegacyHashToCurve` in this case means just set the x coord to the SHA-256 the uncompressed encoding of EC.G, and then i'm not sure about the sign, but should be the same as:
https://github.com/ElementsProject/secp256k1-zkp/blob/1a1a803b7f06032aff6aba5d1de8f574b61e2fcd/src/modules/rangeproof/main_impl.h#L23-L24

`HashToCurve` *could* be the same as `LegacyHashToCurve`, and if that fails when generating the `U` point from `t`, the coordinator can just keep generating random t's until the hash is a valid x coordinate that can be used as `U`, so `LegacyHashToCurve` doesn't even have to be defined for every possible input like a proper hash to curve function.

That said, it does seem to me like a more sensible choice of `HashToCurve` would be a total function, and it makes sense to use the one that libsecp256k1-zkp does IMO (it's also constant time IIRC, but that really doesn't matter for our use case):
https://github.com/ElementsProject/secp256k1-zkp/blob/1a1a803b7f06032aff6aba5d1de8f574b61e2fcd/src/modules/generator/main_impl.h#L70-L211

Reply:
just to be clear I don't feel very strongly about this, also perfectly happy with just hashing the names with a single function... again, the only reason is if people want to implement compatible wabisabi clients using libsecp256k1-zkp for off the shelf Pedersen commitments (and possibly rangeproofs)

Reply:
`Gw`, `Gwp`, `Gx0`... Let me see if I can put this in a diplomatic way: WHAT THE FUCK??! 🤮 🤮 🤮